packages/api/src/services/features.ts

```diff
@@ -0,0 +1,67 @@
+import { Feature } from '../entity/feature'
+import { getRepository } from '../entity/utils'
+import * as jwt from 'jsonwebtoken'
+import { env } from '../env'
+import { IsNull, Not } from 'typeorm'
+
+enum FeatureName {
+  UltraRealisticVoice = 'ultra-realistic-voice',
+}
+
+export const getFeatureName = (name: string): FeatureName | undefined => {
+  return Object.values(FeatureName).find((v) => v === name)
+}
+
+export const optInFeature = async (
+  name: FeatureName,
+  uid: string
+): Promise<Feature | undefined> => {
+  if (name === FeatureName.UltraRealisticVoice) {
+    return optInUltraRealisticVoice(uid)
+  }
+
+  return undefined
+}
+
+const optInUltraRealisticVoice = async (uid: string): Promise<Feature> => {
+  const feature = await getRepository(Feature).findOneBy({
+    user: { id: uid },
+    name: FeatureName.UltraRealisticVoice,
+  })
+  if (feature) {
+    // already opted in
+    console.log('already opted in')
+    return feature
+  }
+
+  // opt in to feature for the first 1000 users
+  const count = await getRepository(Feature).countBy({
+    name: FeatureName.UltraRealisticVoice,
+    grantedAt: Not(IsNull()),
+  })
+
+  let grantedAt: Date | null = new Date()
```

Comment:
Technically there could be a small race condition where more than 1000 could be created, but only a few so I don't think it matters in this case.

Reply:
Yeah, the race condition could happen. I could probably do a `insert` and `select` in a transaction